bindgen/clang.rs

```diff
@@ -555,36 +543,41 @@ impl Cursor {
 
         let (file, other_file) = match (file.name(), other_file.name()) {
             (Some(file), Some(other_file)) => (file, other_file),
-            // Built-in definitions should come first.
-            (Some(_), None) => return cmp::Ordering::Greater,
-            (None, Some(_)) => return cmp::Ordering::Less,
-            (None, None) => return cmp::Ordering::Equal,
+            // Keep the original sorting of built-in definitions.
+            (None, _) | (_, None) => return cmp::Ordering::Equal,
         };
 
+        // If both items are in the same source file, simply compare the offset.
         if file == other_file {
-            // Both items are in the same source file, compare by byte offset.
             return offset.cmp(&other_offset);
         }
 
+        // `None` here means `file`/`other_file` is the main header file.
         let include_location = ctx.included_file_location(&file);
         let other_include_location = ctx.included_file_location(&other_file);
+
         match (include_location, other_include_location) {
-            (Some((file2, offset2)), _) if file2 == other_file => {
+            // The main header file (`None`) comes after header passed as CLI argument (`Some((None, _))`).
+            (None, Some((None, _))) => cmp::Ordering::Greater,
+            (Some((None, _)), None) => cmp::Ordering::Less,
+            // If an item was included in the same source file as the other item,
+            // compare its `#include` location offset the offset of the other item.
+            (Some((Some(file2), offset2)), _) if file2 == other_file => {
                 offset2.cmp(&other_offset)
             }
-            (Some(_), None) => cmp::Ordering::Greater,
-            (_, Some((other_file2, other_offset2))) if file == other_file2 => {
+            (_, Some((Some(other_file2), other_offset2)))
+                if file == other_file2 =>
+            {
                 offset.cmp(&other_offset2)
             }
-            (None, Some(_)) => cmp::Ordering::Less,
-            (Some((file2, offset2)), Some((other_file2, other_offset2))) => {
-                if file2 == other_file2 {
-                    offset2.cmp(&other_offset2)
-                } else {
-                    cmp::Ordering::Equal
-                }
+            // If both items were included in the same file, compare the offset of their `#include` directives.
+            (Some((file2, offset2)), Some((other_file2, other_offset2)))
+                if file2 == other_file2 =>
+            {
+                offset2.cmp(&other_offset2)
             }
-            (None, None) => cmp::Ordering::Equal,
+            // Otherwise, keep the original sorting.
+            _ => cmp::Ordering::Equal,
```

Comment:
I guess this gets the fallback, but is this ok? Seems this could end up with weird cases where the partial ord invariants are not held up and we end up with really weird sorting right?

Reply:
Remove the fallback, all cases should be handled explicitly now.

Reply:
This might be a bit evil but what happens if you do the following:
```h
// a.h
#ifndef A_H
#define A_H

struct foo {};

#include "b.h"

struct baz {
    struct bar field;
};

#endif

```
```h
//b.h
#include "a.h"

struct bar {
  struct foo field;
};
``
Does this ordering guarantee that we don't loop forever because `a.h` is included in `b.h` which includes `a.h`?

Reply:
@pvdrz, yes, we won't loop forever, but there is still an error here.

If `a.h` is the main header file and we compare `baz` with `bar`, we check the include location of `baz`, which is erroneously inside `b.h`, resulting in `bar` wrongly being sorted after `baz`.

The fix for this is easy though: Simply ignore include directives which include the main header.

Reply:
:+1: could you add this as a test?